ranged-based `for` for user-defined types

Pixep opened the pull request:
The goal of this proposal is to provide a way for user-defined types to support range-based iteration with `for`.
The current proposed solution exposes 3 interfaces that can be implemented by user types to enable support for
ranged-for loops.

Co-authored-by: josh11b <josh11b@users.noreply.github.com>

Pixep committed:
Filling out template with PR 1885

Pixep committed:
Add context, other languages, and basic interfaces proposal

Pixep committed:
Cleanup proposal structure, add context and proposal details

zygoloid commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+
+### Consumable
+
+An interface representing single-use sequential access type of container. Items
+ca be read (consumed) once, and are not expected to be accessible anymore after
+usage.
```
Some examples of types that might implement this would be useful to me here. Certainly iterator types could implement this, but it seems a little unusual to treat an iterator as a range. Similarly, destructive traversal over a linked list could implement this (and a destructive traversal might in fact be better than traversal followed by destruction due to improved locality). I could also imagine this being used to good effect when implementing a generator.

Is there an example of a common vocabulary type that you'd expect to implement this?

Pixep replied:
This is a bit different yes, and steps into data ownership territory. The two use cases I had in mind were generators and producer/consumer scenarios, and having an interface to make this explicit.
In terms of common types, this would apply for sure to queues and stacks, as well as any iterable (in the general sense). Generators too, though the semantic added value seems a bit more limited.

From a user standpoint, this could be something like:
```
for (e: auto in queue as Consumable) {
  ...
}
```
Which conveys the intent, independent of the container. In this example I would expect it to just `pop` until empty.
It is definitely a bit more unusual, so I'm not opposed to moving it to a different proposal, changing or dropping it.

And I will add some examples to the description here if this is deemed still valuable.

Pixep replied:
Parked `Consumable` in 00ef294

zygoloid commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+    -   If the `Optional` is null, break the loop. Otherwise return the value
+        held by the Optional.
+
+#### Interface and usage
```
It seems to me that `Iterator` is strictly more general than the other two interfaces, and that we can provide
```
impl forall [T:! Consumable] T as Iterable;
impl forall [T!: Indexable] T as Iterable;
```
in the library. That'd allow us to define the semantics of `for` entirely in terms of `Iterable`, rather than having there be three different interfaces that the library knows about.

Is that possible, or is there some subtlety that I'm missing here? If it is possible, it seems desirable to keep the language <-> library interface simpler.

Pixep replied:
Absolutely, yes. I think there would be a small performance hit due to the indirection. Would that be acceptable? That case that might be a small gotcha of the language for people really performance-focused.

If it is, we might want to be resolve it by just accepting one single interface, and letting users take care of adapting for `Iterable` if necessary. I agree that the simpler the lib interface the better.

zygoloid replied:
I think we should assume that the indirection won't introduce any additional runtime cost in optimizing build modes. And I think there's real value in making the language interface for `for` as simple and minimal as possible, and defining any higher-level abstractions in the library.

Pixep replied:
I parked `Consumable` in 00ef294, and noted possible compatibility with `forall`.

zygoloid commented on proposals/p1885.md:
````diff
@@ -0,0 +1,458 @@
+
+```rust
+// Carbon definition
+interface Indexable() {
````
```suggestion
interface Indexable {
```
You don't need the parens here if the interface has no parameters.

zygoloid reviewed (commented):
Thanks, this is looking promising!

I think we should have at least a sketch of a plan for mutability in this proposal. Given that our main way of providing mutable handles to objects is via pointers, I think we'd want some way to convert a container into a range of pointers to container elements. I think that can be done without any changes to this proposal, for example:
```
fn F() {
  var v: Array(i32, 5) = ...;
  for (p: i32* in v.MutableElements()) {
    // ...
  }
}
```
where `MutableElements` produces an `Iterable where .ElementType == .Self.ElementType*`. (Maybe we'd want a different interface for converting the container to a sequence of element pointers, but we don't need to decide on that here.)

geoffromer commented on docs/design/generics/overview.md:
```diff
@@ -493,7 +493,7 @@ cast from `T` to `CDCover`.
 ### Adapting types
 
-Carbon has a mechanism called [adapting types](terminology.md#adapting-a-type))
+Carbon has a mechanism called [adapting types](terminology.md#adapting-a-type)
```
This should probably be split out as a separate PR, since it's unrelated to the main topic of this PR and can be merged a lot faster.

Pixep replied:
That's fair, created here: https://github.com/carbon-language/carbon-lang/pull/2025

geoffromer commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+-   or B. define adapters or implicit conversions to respectively simplify
+    algorithms implementation or end-user usage
+
+### Indexable
```
Designing an indexing API for Carbon turns out to be surprisingly thorny -- see #1356. It might be better to leave this piece as future work.

Pixep replied:
Thanks for linking that, I missed this proposal. Yes especially as #1356 is not approved / merged yet, this seems backwards to address it here.
Will move to "Alternative considered"

Pixep replied:
`Indexable` was moved to the last section of solutions considered.

geoffromer commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+// Carbon definition
+interface Indexable() {
+    let ElementType:! Type
+    fn GetSize[me: Self]() -> u32;
+    fn GetAt[me: Self](pos: u32) -> ElementType;
```
This would imply that Carbon containers that want to allow more than 2^32 elements can't support indexing. So for example, a C++ `std::vector` couldn't support indexing in Carbon. That seems like an unfortunate restriction. Also, unsigned array indices are prone to bugs, so it seems like this should be `i64`, or maybe the Carbon equivalent of `ptrdiff_t`, if we have such a thing.

In any event, it seems like it'd be a substantial regression from C++ if Carbon's `[]` syntax only supported one particular index type, so I think this type needs to be either a parameter or an associated type, rather than being fixed by the interface. But that raises the question of how to determine the start index value of an arbitrary type, and how to increment it. That's solvable with one or two more interfaces, but I'm not sure you want this proposal to get that far off into the weeds. This seems like another reason to leave this part as future work.

Pixep replied:
That's a good point. The index limitation wasn't intentional when writing this proposal, more an initial placeholder that stayed. For which a parameter or associated type would be the way to go.

Based on the previous comment, will move the entire section to "Alternatives considered"

geoffromer commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+The solution highlighted embeds the end condition in `Next()`, returns a null
+`Optional` at the last elements. This guarantees that `Iterator`s cannot be used
+improperly (code safety goal), and exposes a simple interface, but does have a
+performance cost (cost of the `Optional`).
```
It seems like it could also have an interoperability and migration cost, because Carbon ranged-for loops won't work with C++ containers, and vice-versa.

jonmeow replied:
One option for addressing interoperability without binding ourselves to the C++ API would be to have the interop layer automatically inject adopters for language-conforming iterators. e.g., a Carbon `Iterable` could get adapted to look like a C++ iterator for C++ code, whereas a C++ iterator could have an external impl defined for Carbon code (and likely _should_ need that regardless), letting each language keep its own idioms.

(it may make sense to mark interop as an open question since it's not designed out?)

geoffromer replied:
I think it will be challenging to bridge the gap between these two APIs with adapters, because there's a substantial "impedance mismatch" between them that's hard to hide.

- With this design, a Carbon iterator basically plays the same role as a begin/end iterator _pair_ in C++, so there isn't a one-to-one relationship between the objects on either side of the adaptation. That will be a particular problem for automatic adaptation, but may also cause problems for manual adaptation.
- A C++ forward iterator (or better) must support calling `operator*` multiple times, returning the same result so long as the iterator is not advanced. To satisfy that requirement, an adaptor would have to cache the value returned from `Next`.
- C++ iterators support mutation via `operator*`. That compounds the previous problem, because mutation has to apply to the original element, not a cached copy.
- We could avoid those last two problems by saying that Carbon iterators are exposed in C++ as input iterators, but that excludes mutating use cases, and could dramatically degrade performance even for read-only cases, because many algorithms can't be implemented efficiently (or at all) in terms of the input iterator API.

Pixep replied:
The last points definitely highlight the need for a `Get` function, in place of a simpler `Next`-only approach. I propose to move in that direction.

Regarding the iterator pairs, it is an option to expose an way to get an end iterator for comparison purposes, and interoperability. From a user standpoint, I would probably prefer to not be forced to rely on this end operator, which can be akward at times (needing the container in scope and such), but as a additional feature I believe it makes sense, and likely to have other uses (reverse and such).

**Edit**: Proposal amended with `Get` for `Iterator`.

Pixep committed:
docs: revert typo fix to generics, separate PR

Moved to
https://github.com/carbon-language/carbon-lang/pull/2025

jonmeow commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+
+Interfaces can use
+[type parameter](/docs/design/generics/terminology.md#interface-type-parameters-and-associated-types)
+as `interface Iterator(ItemType:! Type` instead of associated types. This would
```
```suggestion
as `interface Iterator(ItemType:! Type)` instead of associated types. This would
```

geoffromer commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+
+The current `Iterator` interfaces allows simple extension to add features such
+as bidirectional access (adding `Prev()` method), or mutable access
+(`GetMutable[addr me: Self*]() -> ElementType*`, or `Set()`)
```
Should there be a non-mutable `Get` as well? It seems like it would be awkwardly non-orthogonal that if you want to access the element without mutating the iterator, you need to obtain mutable access to the element.

Pixep replied:
Correct. This should probably be `NextMutable` instead, and doesn't make much sense alone. But based on the interoperability conversation, I will go ahead and add `Get`, which will have the benefit of providing a symmetric solution for this.

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+### Consumable
+
+An interface representing single-use sequential access type of container. Items
+ca be read (consumed) once, and are not expected to be accessible anymore after
```
```suggestion
can be read (consumed) once, and are not expected to be accessible anymore after
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+    -   [Other](#other)
+
+<!-- tocstop -->
+
```
Proposals as of #1895 include an abstract:

```suggestion

# Abstract

The goal of this proposal is to provide a way for user-defined types to
support range-based iteration with `for`. The proposed solution exposes
3 interfaces that can be implemented by user types to enable support for
ranged-`for` loops.

```

Pixep replied:
Added

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+    -   `Consumable`
+
+Due to the semantic and functional difference between these interfaces, they
+should not conflict with the [One way](/docs/project/principles/one_way.md).
```
```suggestion
should not conflict with the
 ["one way" principle](/docs/project/principles/one_way.md).
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+
+-   A. explicitly support multiple interfaces: expressive, and specialize
+    implementations based on type features (likely preferred)
+-   or B. define adapters or implicit conversions to respectively simplify
```
```suggestion
    implementations based on type features (likely preferred), or
-   B. define adapters or implicit conversions to respectively simplify
```

josh11b commented on proposals/p1885.md:
````diff
@@ -0,0 +1,458 @@
+
+#### Interface and usage
+
+```rust
````
Carbon code should use ` ```carbon ` even if that isn't supported yet.

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+-   Random access:
+    -   `Indexable`
+-   Sequential access:
+    -   `Iterable`, `Iterator`
```
Having multiple interfaces raises the question of what happens if you implement more than one? Since `for` loops won't ever access the elements except sequentially, it seems better to say something like:

`for` will always use the sequential interfaces. There is also a random access interface. If a type implements the random access interface, an implementation of the sequential interface will be provided automatically, using a `final` blanket `impl`.

Pixep replied:
Right. @zygoloid had a similar comment, I parked `Consumable` in 00ef294, and mentioned possibility for blanket `impl`

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+    -   `Indexable`
+-   Sequential access:
+    -   `Iterable`, `Iterator`
+    -   `Consumable`
```
If we are to support `for` loops that consume the sequence, it seems like there should be some indication in the code that we are consuming (switching to this different interface). This is both valuable as documentation to readers of the code and clarifies what happens if both the sequential and consuming interfaces are implemented for a type.

Pixep replied:
Parked Consumable in https://github.com/carbon-language/carbon-lang/commit/00ef2942718fff07346a7ceed16e424f182a98bf

josh11b commented on proposals/p1885.md:
````diff
@@ -0,0 +1,458 @@
+    let IteratorType:! Iterator;
+    fn GetIterator[me: Self]() -> IteratorType;
+}
+```
````
My understanding is that the leads would prefer the iterable type to be involved in the advancing of the iterator.

Pixep replied:
The `Iterable` type, or the `Iterator`? If we are talking about the `Iterable` (i.e. container), I'm not sure how this is going to work without restricting it to a single global iterator, which does not seem right. Maybe I missed your point?

Pixep replied:
@josh11b is it something you could clarify? Thanks!

josh11b replied:
This is the cursor approach, I've mentioned on discord. All the methods would be on the `Iterable` type, so there would not need to be any `Iterator` interface. Something like:

```
interface Iterable {
  let ElementType:! Type;
  let CursorType:! Type;
  fn NewCursor[me: Self]() -> CursorType;
  fn GetAtCursor[me: Self](cursor: CursorType) -> ElementType;
  fn AdvanceCursor[me: Self](cursor: CursorType*) -> bool;
}
```
Or:
```
interface Iterable {
  let ElementType:! Type;
  let CursorType:! Type;
  fn NewCursor[me: Self]() -> CursorType;
  fn CursorDone[me: Self](cursor: CursorType) -> bool;
  fn GetAtCursor[me: Self](cursor: CursorType) -> ElementType;
  fn AdvanceCursor[me: Self](cursor: CursorType*);
}
```

An advantage of this approach is that an iterator through an array can do bounds checking in debug mode.

Pixep replied:
Switched to cursor

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+-   Compatible with future algorithms (`sort`, `filter`, `transform`, ...)
+-   Suitable for chaining (`std::ranges::views`)
+-   Suitable for lazy evaluation / generators
+
```
Would be good to talk about use cases here. For example, there are lots of containers that have multiple ways of iterating through them:
- forward and reverse
- maps and hash-maps have items, keys, and values

It would also be worth talking about synthetic data like iterating through the numbers 0 to 1,000,000.

Pixep replied:
Done

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+The original `for` proposal does not define a way for user-defined types to
+opt-in to support range-based for loops. An approach is proposed in this
+document.
+
```
Maybe talk about the advantages of the cursor versus iterator approaches. Cursors are things that only make sense in the context of a container, so an integer could be a cursor marking a position in an array. Iterators are things that have all their context, like a pointer marking a position in an array, except possibly when to stop. I have heard Carbon lead @chandlerc say he favors the cursor approach because it makes it easier to have bounds checking in debug mode (and possibly other benefits).

Pixep replied:
Added a small paragraph to that effect. Do you think this challenges the iterator approach in this proposal? I can imagine a situation where this could work provided the "container" implements a specific 'indexable' interface (removed from here, and a WIP by @geoffromer IIRC), that would likely put an unecessary random-access constraint on the said user-container.

geoffromer replied:
I just realized one major advantage of the "cursor" model over the "iterator" model: it's really hard to see how array r-values could have iterators, since r-values don't have addresses, and an iterator is the functional equivalent of a pointer.

Of course, this problem goes beyond `for` loops: it implies we can't run STL algorithms over array r-values, so we might need to solve this some other way anyway. But I think the cursor model needs to at least be in "alternatives considered", if not the primary proposal.

Pixep replied:
Switched to cursor, and iterator now in alternatives. Mentioned R-value simplification.

zygoloid commented on proposals/p1885.md:
````diff
@@ -0,0 +1,458 @@
+
+```rust
+// Carbon definition
+interface Consumable() {
````
````suggestion
```carbon
// Carbon definition
interface Consumable {
````

zygoloid commented on proposals/p1885.md:
````diff
@@ -0,0 +1,458 @@
+
+```rust
+// Basic immutable iterator
+interface Iterator() {
````
````suggestion
```carbon
// Basic immutable iterator
interface Iterator {
````

zygoloid commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+}
+
+// Iterable type
+interface Iterable() {
```
```suggestion
interface Iterable {
```

zygoloid commented on proposals/p1885.md:
```diff
@@ -0,0 +1,458 @@
+// Basic immutable iterator
+interface Iterator() {
+    let ElementType:! Type;
+    fn Next[me: Self]() -> Optional(ElementType);
```
I wonder if there's a performance cost here in using `Optional(ElementType)`. Specifically, for something like `for (var x: LargeThing in sequence)`, do we pay any additional cost for a copy out  of an `Optional(T)` that we might avoid if `Next` returned a `T` directly (with a separate `HasNext` function)? In C++, if `sequence`'s iterators create new `LargeThing`  objects on the fly, that loop would perform no copies, but this iterator design seems to force a copy.

On the other hand, coupling the `Next` and `HasNext` functions seems very nice from an ergonomics and semantics perspective.

Pixep replied:
It is definitely going to boil down to preference from an ergonomics and performance standpoint. If we put an empasis on performance, this might just not be an option. Offering a `SafeIterator` might do it, with the annoying multiplication of interfaces as a side effect (`SafeMutableIterator`?)

Pixep replied:
Removed `Optional` and added to "Alternative considered"

Pixep committed:
Fix various typos

Co-authored-by: josh11b <josh11b@users.noreply.github.com>
Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

Pixep committed:
Drop 'Indexable' interface proposal

Depends on other discussion, and too large for this proposal

Pixep committed:
Add abstract section

Pixep committed:
Fix invalid code flags, `()` for interfaces

Pixep committed:
Add a basic cursor vs iterator paragraph

Pixep committed:
Add `Get` method to `Iterator`

Necessary for C++ interoperability with iterators.

Pixep committed:
Fix typos

Pixep committed:
Drop 'TODO' section

Pixep committed:
Park `Consumable` interface

Consumable interface can be handled in a different proposal, and is not
key to moving forward for the range-based for support.

geoffromer commented on proposals/p1885.md:
```diff
@@ -81,6 +85,12 @@ The original `for` proposal does not define a way for user-defined types to
 opt-in to support range-based for loops. An approach is proposed in this
 document.
 
+This can be achieved using either a cursor or an iterator. A cursor benefits
```
It's probably worth spelling out the difference between "cursor" and "iterator"; I'm not sure everyone is familiar with that terminology.

geoffromer commented on proposals/p1885.md:
```diff
@@ -0,0 +1,417 @@
+interface Iterator {
+    let ElementType:! Type;
+    // Needed in the broader context of Iterators and C++ interop
+    fn Get[me: Self]() -> Optional(ElementType);
+    fn Next[addr me: *Self]() -> Optional(ElementType);
```
Could we define an `Advance` method that moves the iterator forward but doesn't return a value, and then give `Next` a `final` definition in terms of `Advance` and `Get`? That seems more orthogonal and easier to implement, and exposing the ability to increment without the overhead of returning a value could be a useful optimization in some cases.

Pixep replied:
Agree, done

geoffromer commented on proposals/p1885.md:
````diff
@@ -0,0 +1,417 @@
+// `container` is now empty
+```
+
+### Limitations and future language improvements
````
We should probably also mention language support for `const`, and language support for user-defined l-value expressions, because those will determine how iterators should support mutability.

geoffromer commented on proposals/p1885.md:
```diff
@@ -0,0 +1,417 @@
+
+The design of this proposal would benefit from future support for:
+
+-   `Optional` (required)
```
I think we should specifically call out the performance issue of needing to copy the element, either here or as a separate bullet point, because it's a major limitation of this proposal, and an area where we're actively discussing language changes to address it.

Pixep replied:
Done, and updated interface to reomve `Optional`

Pixep committed:
Add a list of use cases

Pixep committed:
Merge branch 'trunk' into proposal-for-statement

Pixep committed:
p1885: add Iterator `Advance()` method

Pixep committed:
p1885: refer to `Advance` in for-loop logic

Pixep committed:
p1885: add mutable iterators

Pixep committed:
p1885: add support suggestions for main use cases

Pixep committed:
p1885: switch from `Optional(T)` to `T` for getters

Pixep committed:
p1885: fix interface name typo

josh11b commented on proposals/p1885.md:
```diff
@@ -22,44 +23,57 @@ SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+The goal of this proposal is to provide a way for user-defined types to support
+range-based iteration with `for`. The proposed solution exposes an interface
+that can be implemented by user types to enable support for ranged-`for` loops.
```
Feels like this could be a little more concise.

```suggestion
This proposes an interface that can be implemented by user types to support
range-based iteration with `for`.
```

josh11b commented on proposals/p1885.md:
```diff
@@ -22,44 +23,57 @@ SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+Example of data and use cases include:
+
+-   Iterating over mutable and immutable data
+-   Iterating over r-value container
```
```suggestion
-   Iterating over an r-value container
```

josh11b commented on proposals/p1885.md:
````diff
@@ -175,111 +195,176 @@ https://stackoverflow.com/questions/35810674/in-go-is-it-possible-to-iterate-ove
+
+```carbon
+// Basic immutable iterator
+interface Iterator {
````
Why go with the iterator approach instead of the cursor approach? I think the cursor approach has a number of advantages, and is what I've heard @chandlerc prefers. For example, I think the cursor approach supports iterating through r-value arrays.

geoffromer replied:
Iterators are likely to interop better with C++. But yes, cursors should definitely at least be an "alternative considered".

IMO we should really avoid statements like "I've heard @chandlerc prefers it".
- It's an argument from authority.
- Citing one lead in that way tends to undermine the other two.
- It's (at least) third-hand information, so there's a high risk of information getting lost, and there's no way to engage with it constructively.

josh11b replied:
I'm sorry, it was not my intention to make an argument from authority. I was trying to say something about how I had heard different stories in different venues from different people and I was hoping they would connect to discuss it. My understanding of the goals of the proposal process is to gather and represent the different opinions and approaches, ideally before the proposal is sent to the leads for final ratification and I perceived a gap. Unfortunately, my attempt to add @chandlerc to the conversation did not lead to that outcome. Next time I will try harder to make it clear that this was an invitation to continue a conversion rather than attempt to end it.

Pixep replied:
Moved to cursor approach following discussion with leads.

Pixep committed:
Apply suggestions from code review

p1885: rewording, fix typo

Co-authored-by: josh11b <josh11b@users.noreply.github.com>

Pixep committed:
p1885: update from google docs

Pixep committed:
p1885: drop outdated alternatives considered

Pixep committed:
p1885: drop reference to future work from "Details"

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,712 @@
+template constraint CppIterator { ... }
+template constraint CppContainer {
+  // speculative
+  forsome IteratorType: CppIterator;
```
Having thought about it more, we need this type to be unique if it is going to be deduced, which is a requirement for using its value (which we do when writing `T.(CppContainer.IteratorType)` below). So it would be better if this speculative syntax used `for_unique` instead of `forsome`:
```suggestion
  for_unique IteratorType: CppIterator;
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,712 @@
+  fn end() -> IteratorType;
+}
+external impl forall [template T:! CppContainer] T as Iterate
+   where  .CursorType = T.(CppContainer.IteratorType)
```
```suggestion
    where .CursorType = T.(CppContainer.IteratorType)
```

Pixep committed:
p1885: fix superfluous line breaks

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,712 @@
+adapter CppIterate(template T:! type) for T {
+  impl as Iterate
+      // speculative
+      where  .CursorType = typeof(T.begin())
```
```suggestion
      where .CursorType = typeof(T.begin())
```

josh11b commented on proposals/p1885.md:
````diff
@@ -0,0 +1,712 @@
+for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
+```
+
+These two options could allow interoperability with compatible C++ type, either
````
```suggestion
These two options could allow interoperability with a compatible C++ type, either
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,712 @@
+    ++*cursor;
+    return Optional(ElementType).MakeValue(value);
+  }
+}
```
```suggestion
}
// used like:
var v: Cpp.std.vector(i32) = ...
for (i: i32 in v) { ... }
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+### Single interface for immutable and mutable data
+
+Because some collections will be read-only, we want to separate mutable and
+immutable interfaces to opt-in only on demand. A single interface would force developers
```
```suggestion
immutable interfaces. A single interface would force developers
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+On the other hand, the cursor approach deviates more from C++ than iterator, and
+may require some more effort for C++ interoperability.
+
+### Single interface for immutable and mutable data
```
Clarify that this is talking about not having methods for handling both immutable and mutable data in one interface. Right now we are talking about having a single interface, and using it with `T` for immutable data, and `T*` for mutable access.

Pixep replied:
Done

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+
+#### C++
+
+C++ requires us to use the very rich
+[`iterator`](https://en.cppreference.com/w/cpp/iterator) library with
+`[c]begin()` and `[c]end()`, to support range-based for.
```
I think the relevant link is: https://en.cppreference.com/w/cpp/language/range-for

The iterator library is not needed. The point is that anything that supports certain methods or function calls may be used.

Pixep replied:
Thank you, that's the reference I was looking for at the time. Updated.

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+#### Python
+
+Python requires 2 dunder methods, `__iter__` on the container and `__next__` on
+the iterator, to be usable with a `for element in container:` construct.
```
Maybe also talk about the generator approach for creating an iterator, since that demonstrates how inversion of control works?

Pixep replied:
Done

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+
+Go has non-standard support for range-based with custom types only.
+
+https://stackoverflow.com/questions/35810674/in-go-is-it-possible-to-iterate-over-a-custom-type
```
I think it would be better to say Go does not officially support customizing `for` for user types (in line with its philosophy of not supporting operator overloading) than refer to this hack.

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+-   Combines "advance", "get", and "bounds check" into a single `Next(cursor)`
+    method that returns an optional value.
+
+The proposed `Iterate` is:
```
Try to write this such that it would read naturally once the proposal is accepted, at which point all of this will be true rather than "proposed."

```suggestion
The `Iterate` interface is:
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+
+The cursor tracks progression, and the `Next` method advances the cursor and
+returns an `Optional` value. An empty `Optional` indicates that we cannot
+advance further.
```
The current wording is a bit unclear to me since it only talks about the "advance" part of `Next` not the "get" part. I'm left not knowing whether there is a final value or not.

```suggestion
returns an `Optional` value. An empty `Optional` indicates that we have reached
the end.
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+      if (*cursor < self.Size()) {
+        let index: i32 = *cursor;
+        *cursor = index + 1;
+        return Optional(i32).Create(self.data[index]);
```
Not entirely sure whether this type checks / is legal Carbon. It would depend on there being a special case for tuples that have equal element types.

Pixep replied:
Agree its definitely not, thanks for highlighting it

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+#### `let` by default
+
+To keep usage simple and non-surprising, we propose defaulting variable
+declarations to `let` declarations.
```
Say something how this matches what we do in function declarations and match statements (see #2188 ).

josh11b commented on proposals/p1885.md:
````diff
@@ -0,0 +1,710 @@
+
+```carbon
+for (var item: T in my_range) {
+  // `item` can be modified, but this will not modify `my_range` because it is a copy of the element with its own storage
````
```suggestion
  // `item` can be modified, but this will not modify `my_range`
  // because it is a copy of the element with its own storage.
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+
+R-values are likely to be limited in terms of addressing and mutation. The
+impact is undefined until the
+[corresponding design](https://github.com/carbon-language/carbon-lang/pull/2006/files)
```
```suggestion
[corresponding design](https://github.com/carbon-language/carbon-lang/pull/2006)
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+
+On the other hand, changing how the collection is traversed (for example
+reverse), requires internal knowledge about the data layout, and a custom
+implementation. One approach for containers that support reverse iteration would
+be to implement a standard `IterateBidirectional` interface with a `Prev()`
+method. A proxy could be exposed to make it usable with `for` directly.
```
Is the `IterateBidirectional` interface part of the proposal or future work?

Pixep replied:
Clarified this is an **example**

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+
+### C++ interoperability
+
+The limitation of this approach is the need to adapt to C++ iterator model. This
```
```suggestion
The limitation of this approach is the need to interoperate with the C++ iterator model. This
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+### C++ interoperability
+
+The limitation of this approach is the need to adapt to C++ iterator model. This
+could be implemented by having the cursor be an iterator type:
```
There are two directions for interop. I think what you are saying is that we can create a Carbon cursor from a C++ iterator, but I think that should be clarified. It raises the question about interop in the other direction, which should ideally be addressed as well. Maybe there should be two subsections to make it very explicit?

josh11b commented on proposals/p1885.md:
````diff
@@ -0,0 +1,710 @@
+
+```carbon
+// adapter approach
+adapter CppIterate(template T:! type) for T {
````
May be worth noting that this approach could have a cursor type parameter (with some sort of default computed from the return type of `T.begin()`), avoiding the deduction issues of the above approach.

Pixep replied:
Added a mention to that effect.

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+
+### Inversion of control
+
+Using an Optional has limitations, discussed in the
```
```suggestion
Using an `Optional` has downsides, discussed in the
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+
+Using an Optional has limitations, discussed in the
+["alternatives considered"](#alternatives-considered) section. A possible
+approach would be to invert the control of the for loop, by having the container
```
```suggestion
approach would be to invert the control of the `for` loop, by having the container
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+
+This has the following advantages:
+
+-   Removes the need for an Optional, associated overheads, copies, and
```
```suggestion
-   Removes the need for an `Optional`, and the associated overheads, copies, and
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+Using a single operation to return an optional is also consistent with Rust’s
+approach.
+
+The drawback of this `Next` method is the fact it returns an `Optional`, though
```
This paragraph would be clearer if it stated what the downsides of returning an optional was.

Pixep replied:
Changed the way this is presented, does this look better to you?

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+ways to limit or eliminate that cost seem to exist (value reuse or view for
+Optional, inversion of control on `for` block statement).
+
+### Returning `T` instead of `Optional(T)`
+
+Using `Optional(T)` is a side effect of combining operations, discussed in the
+previous section.
+
```
I don't think this section is adding any additional value beyond the previous section.

```suggestion
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,710 @@
+This proposal offers a first step to support a needed way to iterate using a
+`for` loop. More specifically:
+
+-   Requires a single interface to be implemented, uses cursors instead of their
```
```suggestion
-   Requires a single interface to be implemented with a combined `Next` method,
    uses cursors instead of their
```

Pixep committed:
p1885: wording improvements and precisions from josh11b

Co-authored-by: josh11b <josh11b@users.noreply.github.com>

Pixep committed:
p1885: more changes

Pixep committed:
p1885: more small changes

Pixep committed:
p1885: clarified interface example

Pixep committed:
p1885: expand on comments

Pixep committed:
p1885: add python `yield` example

josh11b commented on proposals/p1885.md:
```diff
@@ -100,14 +99,29 @@ user-defined types is addressed in different languages.
+the iterator, to be usable with a `for <element> in <container>:` construct.
+
+A lazy iterator can also be returned using `yield`, giving directly control over
+the values presented for each iteration.
```
```suggestion
[Generator functions](https://peps.python.org/pep-0255/) act like iterators.
They are executed until the next `yield` statement, whose argument is returned
as the next value to the `for` loop. Iteration ends when the function returns.
```

josh11b commented on proposals/p1885.md:
````diff
@@ -100,14 +99,29 @@ user-defined types is addressed in different languages.
+the values presented for each iteration.
+
+```python
+def generate():
````
```suggestion
def generate_fibonacci():
```

josh11b commented on proposals/p1885.md:
```diff
@@ -625,25 +648,25 @@ adapter CppIterate(template T:! type) for T {
-approach would be to invert the control of the for loop, by having the container
-call in the `for` block for each value, passing the value as argument.
+approach would be to invert the control of the `for` loop, by having the
+container call in the `for` block for each value, passing the value as argument.
```
```suggestion
container call in the `for` block for each value, passing the value as argument.
This would work similarly to [Python generator functions](#python).
```

josh11b commented on proposals/p1885.md:
```diff
@@ -678,14 +701,15 @@ together, providing more opportunities for optimization.
-The drawback of this `Next` method is the fact it returns an `Optional`, though
-ways to limit or eliminate that cost seem to exist (value reuse or view for
-Optional, inversion of control on `for` block statement).
+The combined `Next` method forces us to return an `Optional`, which has a
```
```suggestion
The combined `Next` method forces us to return an `Optional`, which has
```

josh11b commented on proposals/p1885.md:
```diff
@@ -678,14 +701,15 @@ together, providing more opportunities for optimization.
-Using `Optional(T)` is a side effect of combining operations, discussed in the
-previous section.
+Those may be mitigated by leveraging unused bit patterns, passing a view of the
+value, or inverting control for of `for` block statements.
```
```suggestion
value, or [inverting control](#inversion-of-control).
```

josh11b commented on proposals/p1885.md:
```diff
@@ -678,14 +701,15 @@ together, providing more opportunities for optimization.
+downsides:
 
-### Returning `T` instead of `Optional(T)`
+-   Potential performance overhead to wrapping and unwrapping, and storage
+    overhead for `Optional` itself
+-   May require copying the value wrapped
```
```suggestion
-   potential performance overhead to wrapping and unwrapping,
-   storage overhead for `Optional` itself, and
-   may require additional copying of the value wrapped.
```

josh11b reviewed (changes requested):
Looking good! I think this is pretty close. The only thing I see beyond the suggestions I'm making now is the unresolved comment about the C++ interop section.

Pixep committed:
p1885: more updates

Co-authored-by: josh11b <josh11b@users.noreply.github.com>

aviRon012 commented:
line 253 it says: `let curser : ... `,
if I understand correctly the curser changes during iteration, so shouldn't it be `var curser : ...`?

aviRon012 commented on proposals/p1885.md:
````diff
@@ -0,0 +1,735 @@
+A naive Carbon implementation of the for loop could be:
+
+```carbon
+let cursor: range.(Iterate.CursorType) = range.(Iterate.NewCursor)();
````
if the curser is mutated during the iteration, shouldn't it be `var curser : ...`?

josh11b replied:
You are correct. Good catch!

```suggestion
var cursor: range.(Iterate.CursorType) = range.(Iterate.NewCursor)();
```

Pixep replied:
Thanks!

Pixep committed:
Fix code snippet typo

Co-authored-by: josh11b <josh11b@users.noreply.github.com>

Pixep committed:
Clarified C++ interoperability goals

Pixep committed:
Add carbon to cpp interop section

Pixep commented:
> Looking good! I think this is pretty close. The only thing I see beyond the suggestions I'm making now is the unresolved comment about the C++ interop section.

Updated the interop section in two parts.

geoffromer commented on proposals/p1885.md:
```diff
@@ -0,0 +1,779 @@
+These two options could allow interoperability with a compatible C++ type,
+either implicitly (template approach) or explicitly (adapter approach).
+
+#### Supporting Carbon `Iterate` in C++
```
Does this mean "Implementing the `Iterate` interface in C++ code", or "Implementing C++ iteration in terms of `Iterate`"? The phrasing sounds more like the first, but from context I think you mean the second.

Maybe the previous section should be called "Iterating over C++ types in Carbon", and this section should be called "Iterating over Carbon types in C++".

Pixep replied:
Agree that your suggestions are clearer, thanks!

geoffromer commented on proposals/p1885.md:
```diff
@@ -0,0 +1,779 @@
+
+template <typename C, typename T>
+class CppIterate {
+  C start() { return C::FromCarbonCursor(iterate_.NewCursor()); }
```
```suggestion
  C begin() { return C::FromCarbonCursor(iterate_.NewCursor()); }
```

geoffromer commented on proposals/p1885.md:
````diff
@@ -0,0 +1,779 @@
+
+Below is an incomplete example of classes exposing those methods:
+
+```cpp
````
I'm having difficulty following this example -- is `Cursor` a cursor, or an iterator that wraps a cursor? How does `Cursor` relate to `CppIterate`? What do the template parameters of `CppIterate` mean?

Apart from those API questions, I think it would be very helpful to flesh out the implementation of `Cursor` in more detail, showing the data members and at least some of the method bodies. That would help answer questions like:

- How do we support calling `operator*` twice on the same `Cursor`?
- Assuming that `Cursor` is meant to be a C++ iterator, how do we support copying it?

Pixep replied:
Added a complete (working) implementation. Might not be the best way to do it, but this gives a first idea of what is necessary.

geoffromer commented on proposals/p1885.md:
```diff
@@ -0,0 +1,779 @@
+Large elements would have a negative impact if the `Optional` cannot leverage
+unused bit patterns within the type, nor present a view instead of a copy.
+
+Such opportunities for `Optional` exist, and could be addressed in the future.
```
I'm skeptical that there are opportunities for `Optional(T)` to present a view of the `T` object instead of a copy. Currently, Carbon has no way of expressing that, and there is substantial resistance to adding a way to do so, because a view of an arbitrary type is basically a reference. As far as I can tell, if we want to solve this problem without introducing references, we have to do it as part of the design of `for`, not the design of `Optional`. That way we can define the reference-like behavior as part of the hidden machinery that `for` rewrites to, without giving user code access to the full power of reference types. See #2274 for an example of something similar, where we enabled a particular kind of reference-like behavior by defining special rules for how `[]` is rewritten to a method call.

In many ways inversion of control would be an ideal solution here; it's really unfortunate that it's apparently so hard to optimize.

Pixep replied:
If I understand that correctly, you'd suggest having a similar rewrite that bypasses `Optional`s entirely. Which means needing to expose an `Optional`-less API (but then losing the advantage of having a single `Next` method), or short-circuiting the `return` before the Optional is created (is that possible?). I'm not sure the same situation applies exactly, feel free to correct me.

I was skeptical too, @zygoloid seemed somewhat confident about it in a previous discussion. But I haven't done a good job at capturing here what he had in mind. @zygoloid would you mind sharing again what you had in mind if you recall?

geoffromer replied:
I'm not suggesting bypassing `Optional` entirely; I'm only saying that I don't expect `Optional` to solve the problem of providing view-like performance, and so it will need to be solved as part of the design of `for`.

We could solve this by having `Next()` return `Optional(T*)`, except that doesn't work on its own because the elements of the underlying range might not be lvalues (either because the range itself is an rvalue or because the range type does something like compute elements on the fly). So sometimes we want a next-element operation that returns `Optional(T)`, and other times we want one that returns `Optional(T*)`. That all seems very similar to the problem that we had to solve with subscripting, where sometimes we want a get-nth-element operation that returns `T`, and other times we want one that returns `T*`, so it seems like a similar solution might work here. I think this would also let us support mutating the elements being iterated over, which is a nice bonus.

geoffromer replied:
We discussed this a bit at the weekly meeting today, and a couple of other options came up:
- We could let the container choose whether it exposes values or pointers to values, and maybe provide a dereferencing adapter that turns a range of pointers into a range of values. The nice thing is this doesn't involve any technical changes to the design, just a shift in expectations about how containers will typically implement `Iterate`. However, that seems pretty awkward from a user experience point of view, and I think it would mean that any given container type could support efficient iteration, or iteration over container r-values, but not both, so I think this only makes sense as a stopgap while we investigate a longer-term solution like control inversion.
- We could allow the binding in the `for` loop be an `addr` pattern, which would be implemented by calling into a separate interface on the container. This puts the choice completely in the hands of the `for` loop author, so it wouldn't let us match the ergonomics of `let` (where the compiler can choose to copy or alias the value depending on what's more efficient), but it at least gives us parity with C++ (where the author has to choose whether the iteration variable is a reference).

Pixep replied:
Thank you, updated to reflect the options suggested here, and noted that the preference today seems to be with #1, to get to a first step before revising.

Pixep committed:
Fix iterator member typo

Co-authored-by: Geoff Romer <gromer@google.com>

Pixep committed:
Clarify sections for C++ interop

Pixep committed:
Merge branch 'carbon-language:trunk' into proposal-for-statement

Pixep committed:
Add mitigation for container with large values

Pixep committed:
Add Iterate to C++ interop sample implementation

Pixep committed:
add copy constructor for cursor iterator

geoffromer commented on proposals/p1885.md:
```diff
@@ -668,36 +701,115 @@ for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
+
+namespace Carbon {
+
+template <typename C, typename T>
+class Iterable;
```
Does this forward declaration get used anywhere in `IterateIterator`?

Pixep replied:
Not needed anymore

geoffromer commented on proposals/p1885.md:
```diff
@@ -668,36 +701,115 @@ for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
+template <typename C, typename T>
+class Iterable;
+
+template <typename Iterable>
```
I'm guessing that the template argument can be the C++ counterpart of any Carbon type that implements `Iterate`, right? If so, calling it `Iterable` instead of `Iterate` is confusing. Same goes for the `Iterable` template, and the `MyIterable` class.

Also, I think a comment would make this a lot clearer, e.g.

```suggestion
// A C++ input iterator for any type that implements the Carbon `Iterate` interface.
template <typename Iterable>
```

Pixep replied:
Right, I battled that a minute, intending to use `Iterable` as a kind of concept, rather than the actual `Iterate`. Reads better, but agree its a bit more confusing.

geoffromer commented on proposals/p1885.md:
```diff
@@ -668,36 +701,115 @@ for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
+    return std::nullopt;
+  }
+};
+};  // namespace Carbon
```
```suggestion
}  // namespace Carbon
```

geoffromer commented on proposals/p1885.md:
```diff
@@ -668,36 +701,115 @@ for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
-  C start() { return C::FromCarbonCursor(iterate_.NewCursor()); }
-  C end() { return C::Invalid(); }
-  // [...]
+class Iterable {
+ public:
+  using CursorType = C;
+  using ElementType = T;
+
+  virtual auto NewCursor() const -> CursorType = 0;
+  virtual auto Next(CursorType& cursor) const -> std::optional<ElementType> = 0;
+
+  auto begin() -> IterateIterator<Iterable<C, T>> {
+    return IterateIterator<Iterable<C, T>>(*this, NewCursor());
+  }
+  auto end() -> IterateIterator<Iterable<C, T>> {
+    return IterateIterator<Iterable<C, T>>::Invalid(*this);
+  }
+};
+
+class MyIterable : public Iterable<int, std::string> {
+  auto NewCursor() const -> CursorType override { return 0; }
+  auto Next(CursorType& cursor) const -> std::optional<ElementType> override {
+    if (cursor < 10) {
+      return std::to_string(cursor++);
+    }
+    return std::nullopt;
+  }
+};
```
I think it might be clearer to omit `Iterable` and `MyIterable`, and instead have text saying that the generated C++ counterpart for any Carbon type that implements `Iterate` will have `begin` and `end` methods defined as follows, and then show those methods in isolation. The rest of this code doesn't really address the key issues for this proposal, and raises other questions that aren't important here (e.g. will we really use inheritance?)

Pixep replied:
Great suggestion, thanks.

geoffromer commented on proposals/p1885.md:
```diff
@@ -668,36 +701,115 @@ for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
-Below is an incomplete example of classes exposing those methods:
+Below is sample implementation to iterate on a Carbon `Iterate`. Note that we
+need to cache the last value, to allow for `*it` to be called, and to be called
+more than once.
```
It seems worth calling out that this will mean that the iterator is not copyable if the value type is not copyable. That's not a problem as far as C++'s ranged-for is concerned, but it's a problem for the C++ standard library, which requires all iterators to be copyable.

Pixep replied:
This should be clearer now

Pixep committed:
Update proposals/p1885.md

Co-authored-by: Geoff Romer <gromer@google.com>

Pixep committed:
Improve C++ iterator sample

Pixep committed:
Precise the impact of caching last value

josh11b commented on proposals/p1885.md:
````diff
@@ -582,13 +584,55 @@ for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
+solution, a better alternative will be needed in the future.
+
+```carbon
+adapter IterateByValue(template T:! type) for T {
````
I think this can be made into a checked generic instead of a template if we can find the right constraint:

```suggestion
adapter IterateByValue(T:! Iterate where .CursorType impls Deref) for T {
```

Pixep replied:
Done, I cannot confirm if this syntax is correct, but can rely on others to mention it if needed.

josh11b commented on proposals/p1885.md:
````diff
@@ -582,13 +584,55 @@ for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
+
+```carbon
+adapter IterateByValue(template T:! type) for T {
+  impl as Iterate where
+      // Speculative syntax. Alternatively, a cursor type parameter
+      // would avoid the need for deduction.
+      where .CursorType = T.CursorType
+        and .ElementType = T.CursorType.(Deref.Result) {
````
With the checked generic parameter, I don't think there is anything speculative here, nor anything calling for a cursor type parameter.
```suggestion
  impl as Iterate
      where .CursorType = T.CursorType
        and .ElementType = T.CursorType.(Deref.Result) {
```

Pixep replied:
My bad, though the `Deref` syntax was, but seems like it exists in the doc. Done.

josh11b commented on proposals/p1885.md:
```diff
@@ -582,13 +584,55 @@ for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
+For reference, range-based `for` loops in C++ requires:
+
+-   `begin()` and `end()` methods or free functions
+-   supporting pre-increment operator `++` and indirection operator `*`
```
Two things:
- Something to connect these two requirements to make it clear that the type returned by `begin()` and `end()` is what has to support these operators.
- I believe the iterator type must also support inequality comparison (at least). According to https://en.cppreference.com/w/cpp/language/range-for a C++ range-based `for` loop gets rewritten to include `for ( ; __begin != __end; ++__begin)`.

We may need to be careful about closely matching C++ so we can be as compatible as we can manage, as long as it doesn't impact the usability for ordinary users. In C++, my understanding is as long as the rewrite type checks, the ranged-based for loop is legal, like a template. This means we may need to support `begin()` and `end()` returning different types, as long as `!=` is defined on those two types.

Pixep replied:
Absolutely yes, clarified.

josh11b commented on proposals/p1885.md:
```diff
@@ -582,13 +584,55 @@ for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
+
+See section 8.6.4 of N4860 for more details.
+
+A limitation of the approach proposed in this document is the need for
+adaptation to interoperate with the C++ iterator model. This could be
+implemented by having the cursor be an iterator type:
```
```suggestion
A limitation of the approach proposed in this document is the need for
adaptation to interoperate with the C++ iterator model. The adapter would
be a Carbon type that satisfies the Cursor interface, implemented in terms
of a C++ iterator.
```

josh11b commented on proposals/p1885.md:
```diff
@@ -582,13 +584,55 @@ for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
+adaptation to interoperate with the C++ iterator model. This could be
+implemented by having the cursor be an iterator type:
 
 -   `NewCursor` providing the "start iterator"
```
```suggestion
-   `NewCursor` providing the "start iterator" returned by `begin()`
```

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,864 @@
+Two different approaches are identified:
+
+-   One is to have a templated `impl` of `Iterate`, for anything with the right
+    method signatures ("begin", "end", etc.)
```
```suggestion
    method signatures (`begin`, `end`, and so on)
```

josh11b commented on proposals/p1885.md:
```diff
@@ -657,6 +701,91 @@ for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
 
+#### Iterating over Carbon types in C++
+
+C++ needs to provide `begin()` and `end()` methods or free functions, in
+addition to `*` and `++` operators.
```
```suggestion
There need to be `begin()` and `end()` methods or free functions accessible to
C++ for any Carbon type that implements the `Iterate` interface, in
addition to overloads of the `*`, `++`, and `!=` operators for the types returned.
```

Pixep replied:
Great, thanks for clarifying.

josh11b commented on proposals/p1885.md:
```diff
@@ -657,6 +701,91 @@ for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
+    assert(cursor_);
+    return value_;
+  }
+  auto operator++() -> IterateIterator<Iterate>& {
```
Probably should have both this and the version that takes an `int` parameter, so this iterator supports both pre- and post- increment.

geoffromer replied:
I don't think I agree. Post-increment won't work if `ElementType` isn't copyable. By the same token, this type won't be a valid iterator if `ElementType` isn't copyable, but it will still work with range-based for loops, so I think we should focus solely on the features needed to support that.

Pixep replied:
Agree with @geoffromer , resolving this one as I don't think this would be a blocking issue. Feel free to re-open if you think otherwise @josh11b .

josh11b commented on proposals/p1885.md:
```diff
@@ -657,6 +701,91 @@ for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
+// ElementType `T` could have `begin()` and `end()` methods
+// defined as such:
+class IterateType ... {
+  auto begin() -> IterateIterator<Iterable<C, T>> {
+    return IterateIterator<Iterable<C, T>>(*this, NewCursor());
+  }
+  auto end() -> IterateIterator<Iterable<C, T>> {
+    return IterateIterator<Iterable<C, T>>::Invalid(*this);
```
What is `Iterable<...>` here? It looks like `IterateIterator` expects to be parameterized by a container type. Is `Iterable` some sort of adapter to define the `CursorType` and `ElementType` members? I don't see it defined anywhere else in this file.

Pixep replied:
That's a typo, and should be  `Iterate<C, T>`, which is the C++ equivalent to Carbon's type. Fixed.

geoffromer commented on proposals/p1885.md:
```diff
@@ -712,29 +712,23 @@ invalid value, to satisfy iterator comparison when `Next()` returns an empty
 Below is sample implementation to iterate on a Carbon `Iterate`. Note that we
-need to cache the last value, to allow for `*it` to be called, and to be called
-more than once.
+need to copy and cache the last value, to allow for `*it` to be called. This
+puts a requirement on `ElementType` to be copyable for this interoperability to
+be usable.
```
I think this will still work if `ElementType` is move-only (but see the comment below), it just means that `IterateIterator` will be move-only in those cases. That means it won't be a valid iterator according to the standard library, but it should still work with range-based `for` loops.

Pixep replied:
I just looked back at the code: 
```
     if (const auto v = container_->Next(*cursor_); v) {
      value_ = *std::move(v);
    }
```
Won't `Next()` require `ElementType` to be copyable nonetheless, as we are effectively returning a copy of an element, and not moving anything from the container? Even if we avoid an optional-like wrapper, or just move from it, the `ElementType` still has to be copied at some point (likely withing `Next()`).

Maybe I forgot something between now and then.

geoffromer replied:
In recent versions of C++, if `Next` returns an `ElementType` rather than a reference, the code you quote won't actually copy or move the return value; it will use `v` itself as the storage for the returned temporary. Now, you're right that in this context, the _body_ of `Next` will very likely need to make a copy, but `Next` is written in pure Carbon, so that problem isn't caused by caching in the interoperability layer, and in fact it's not caused by the interop layer at all -- the same problem exists in pure Carbon.

Pixep replied:
Agree that this is by design, and not caused by the interop. To expand, I think we could say "move-only" in the C++ domain, most that would be ignoring a very likely "copiable" requirement on Carbon side / `Next`. So all-in-all, "copyable" seem more correct, and the upper bound of reqs, vs the lower bound.

geoffromer replied:
The problem is that this doc doesn't explicitly talk about copyability on the Carbon side, so by introducing it here as a new topic that's specifically tied to caching in the interop layer, it really makes it sound like native Carbon will not require copyability, but C++ interop will.

I'd suggest revising the "Large elements and Optional" section to explicitly talk about the fact that this design may force `Next` to make a copy, and hence require `ElementType` to be copyable. Then down here you can say that the caching requires `ElementType` to be movable, and note that this won't matter in practice if Carbon requires it to be copyable anyway.

geoffromer commented on proposals/p1885.md:
```diff
@@ -744,73 +738,52 @@ class IterateIterator {
-    if (const auto val = container_->Next(*cursor_); val) {
-      value_ = *val;
+    if (const auto v = container_->Next(*cursor_); v) {
+      value_ = *v;
```
```suggestion
      value_ = *std::move(v);
```

This will enable the code to work with move-only types.

Pixep replied:
Good point, thanks.

Pixep committed:
Add suggestions from @josh11b

Co-authored-by: josh11b <josh11b@users.noreply.github.com>

Pixep committed:
Add suggestion from @geoffromer

Co-authored-by: Geoff Romer <gromer@google.com>

geoffromer commented on proposals/p1885.md:
```diff
@@ -0,0 +1,864 @@
+This proposal offers a first step to support a needed way to iterate using a
+`for` loop. More specifically:
+
+-   Requires a single interface to be implemented with a combined `Next` method,
```
It might also be worth mentioning that a combined `Next` method makes it easier to support ranges that consume their input. For example, in C++20 if you filter an `istream_view` through `take(5)` (which filters a range to its first 5 elements), the first *6* elements of the stream are consumed, because the C++ iterator model forces `istream_iterator` to consume an element on initialization (so that the first `operator*` has a result to return) and on every `operator++` call. As another example of the same kind of problem, `iota(0) | filter([](auto x) { return x < 5; })) | take(5)` is an infinite loop. I think a combined `Next` method will avoid these problems.

I got this from a talk by Barry Rezvin at C++Now last week; I'm not sure when the video will be online, but you can see the slides [here](https://schedule.cppnow.org/wp-content/uploads/2023/03/take5-slides.pdf)

Pixep replied:
Added to the "Alternatives" section, as this seemed to best belong there.

Pixep committed:
Fix typo on `Iterate` class name in c++ snippet

Pixep committed:
Clarified C++ requirements, and implementation example

josh11b commented on proposals/p1885.md:
````diff
@@ -0,0 +1,866 @@
+for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
+```
+
+### Large elements and Optional
````
@chandlerc I feel like we had a conversation about this and r-value representations. Could you recall where we ended up with this?

josh11b commented on proposals/p1885.md:
```diff
@@ -0,0 +1,866 @@
+-   the type returned supports pre-increment `++`, indirection `*`, and
+    inequality `!=` operations
+
+See section 8.6.4 of N4860 for more details.
```
Can this be made into a link?

Pixep replied:
Do we have any preferred source to link the standard in an easy to read way? This is what I have https://eel.is/c++draft/stmt.iter#stmt.ranged from the latest draft.

josh11b replied:
I don't have a preferred source. The link you mention is fine with me. For non-spec C++ links, I tend to use cppreference.com

Pixep replied:
Done

josh11b reviewed (approved):
I think this is about ready for leads review

josh11b reviewed (approved):
I think this is about ready for leads review

Pixep committed:
Link range-based for statement from C++ standard.

Pixep committed:
Mention ranges consuming their input

From Barry Revzin's "take(5)" presentation

zygoloid reviewed (approved):
I'm happy with this. It seems likely to me that we'll want to revisit some aspects of this eventually, such as ergonomic improvements for mutable iteration, but this seems like a good foundation to use as the basis for further exploration.

zygoloid reviewed (approved):
Also approving on behalf of leads. Thank you for working on this!

chandlerc committed:
Force checks to re-run.
